c/src/atom.rs

```diff
@@ -272,7 +310,25 @@ impl Grounded for CGrounded {
     }
 
     fn match_(&self, other: &Atom) -> matcher::MatchResultIter {
-        match_by_equality(self, other)
+        match self.api().match_ {
+            Some(func) => {
+                let mut vec_temp: Vec<Bindings> = Vec::new();
+                let mut bnd_temp = Bindings::new();
+                let var_temp = VariableAtom::new("var_temp");
+                let atom_temp = Atom::sym("atom_temp");
+                bnd_temp.insert(var_temp.clone(), atom_temp.clone());
+                vec_temp.push(bnd_temp);
+                let mut context: *mut c_void = (&mut vec_temp as *mut Vec<Bindings>).cast::<c_void>();
+
+                func(self.get_ptr(), (other as *const Atom).cast::<atom_t>(), CGrounded::match_callback, context);
+
+                let res_vec = unsafe{ &mut *context.cast::<Vec<Bindings>>() };
+                let len_ = res_vec.len();
+                let res_bnd = res_vec.get(len_-1).unwrap();
+                Box::new(std::iter::once(res_bnd.clone()))
+            },
+            None => match_by_equality(self, other)
+        }
```

Comment:
1. `mut` is not needed at line 321, and compiler warns about it
2. Pushing something into the `vec_temp` doesn't make sense, but I believe it is test code, not sure why it is not inside callback or even C caller
3. It also no need to convert the vector again from `context` and returning the results code looks incorrect, what is required is just returning iterator over `vec_temp` content
4. I would also rename `vec_temp` to `results` or something more meaningful

Overall function should look like this:
```Rust
    fn match_(&self, other: &Atom) -> matcher::MatchResultIter {
        match self.api().match_ {
            Some(func) => {
                let mut results: Vec<Bindings> = Vec::new();
                let context: *mut c_void = (&mut results as *mut Vec<Bindings>).cast::<c_void>();
                func(self.get_ptr(), (other as *const Atom).cast::<atom_t>(), CGrounded::match_callback, context);
                Box::new(results.into_iter())
            },
            None => match_by_equality(self, other)
        }
    }
```